src/map_ref.rs

```diff
@@ -0,0 +1,214 @@
+use crate::iter::*;
+use crate::HashMap;
+use crossbeam_epoch::{self as epoch, Guard};
+use std::borrow::Borrow;
+use std::collections::hash_map::RandomState;
+use std::fmt::{self, Debug, Formatter};
+use std::hash::{BuildHasher, Hash};
+use std::ops::Index;
+
+/// A reference to a [`HashMap`], constructed with [`HashMap::as_ref`].
+/// The current thread will be pinned for the duration of this reference.
+pub struct HashMapRef<'map, K, V, S = RandomState> {
+    map: &'map HashMap<K, V, S>,
+    guard: Guard,
+}
+
+impl<K, V, S> HashMap<K, V, S> {
+    /// Get a reference to this map with the current thread pinned.
```

Comment:
Let's also point people to the type-level docs for `HashMapRef` here.

Reply:
Can't they just click the link on the return value?